test/libsolidity/analysis/FunctionCallGraph.cpp

```diff
@@ -0,0 +1,1989 @@
+
+	map<string, EdgeNames> expectedCreationEdges = {
+		{"C", {
+			{"Entry", "constructor of C"},
+		}},
+		{"D", {
+			{"Entry", "constructor of C"},
+			{"constructor of C", "constructor of D"},
+			{"constructor of C", "function free()"},
+		}},
+		{"E", {
+			{"Entry", "constructor of E"},
+			{"Entry", "function free()"},
+		}},
+		{"F", {
+			{"Entry", "constructor of C"},
+			{"constructor of C", "constructor of D"},
+			{"constructor of C", "function free()"},
+			{"constructor of D", "constructor of E"},
+			{"constructor of D", "function free()"},
+			{"constructor of E", "function C.f()"},
+			{"constructor of E", "function free()"},
+			{"constructor of E", "constructor of F"},
+			{"constructor of F", "function C.f()"},
+			{"constructor of F", "function free()"},
+		}},
+	};
+
+	map<string, EdgeNames> expectedDeploymentEdges = {
+		{"C", {}},
+		{"D", {}},
+		{"E", {}},
+		{"F", {}},
+	};
+
+	checkCallGraphExpectations(get<0>(graphs), expectedCreationEdges);
+	checkCallGraphExpectations(get<1>(graphs), expectedDeploymentEdges);
+}
+
+BOOST_AUTO_TEST_CASE(diamond_inheritance)
+{
+	unique_ptr<CompilerStack> compilerStack = parseAndAnalyzeContracts(R"(
+		function fD() returns (uint) {}
+		function fE() returns (uint) {}
+		function fFD() returns (uint) {}
+		function fFE() returns (uint) {}
+
+		function fVarC() returns (uint) {}
+		function fVarD() returns (uint) {}
+		function fVarE() returns (uint) {}
+		function fVarF() returns (uint) {}
+
+		contract C {
+			uint c = fVarC();
+
+			constructor (uint) {}
+		}
+
+		contract D is C(fD()) {
+			uint d = fVarD();
+
+			constructor (uint) {}
+		}
+
+		abstract contract E is C {
+			uint e = fVarE();
+
+			constructor (uint) {}
+		}
+
+		contract F is D(fFD()), E {
+			uint f = fVarF();
+
+			constructor (uint) E(fFE()) {}
+		}
+	)"s);
+	tuple<CallGraphMap, CallGraphMap> graphs = collectGraphs(*compilerStack);
+
+	map<string, EdgeNames> expectedCreationEdges = {
+		{"C", {
+			{"Entry", "constructor of C"},
+			{"Entry", "function fVarC()"},
+		}},
+		{"D", {
+			{"Entry", "constructor of C"},
+			{"Entry", "function fVarC()"},
+			{"constructor of C", "constructor of D"},
+			{"constructor of C", "function fD()"},
+			{"constructor of C", "function fVarD()"},
+		}},
+		{"E", {
+			{"Entry", "constructor of C"},
+			{"Entry", "function fVarC()"},
+			{"constructor of C", "constructor of E"},
+			{"constructor of C", "function fVarE()"},
+		}},
+		{"F", {
+			{"Entry", "constructor of C"},
+			{"Entry", "function fVarC()"},
+			{"constructor of C", "constructor of D"},
+			{"constructor of C", "function fD()"},
+			{"constructor of C", "function fVarD()"},
+			{"constructor of D", "constructor of E"},
+			{"constructor of D", "function fVarE()"},
+			{"constructor of E", "constructor of F"},
+			{"constructor of E", "function fFD()"},
+			{"constructor of E", "function fVarF()"},
+			{"constructor of F", "function fFE()"},
+		}},
+	};
```

Comment:
I'm not sure how the order in which constructors call each other (and functions called from state variable initializers, inheritance specifiers and constructor modifier list) should look like. This probably needs some discussion so I'll only point out some things that stand out here:
- Functions called from state variable initializers currently are marked as called by the constructor one level below (or `Entry` when we're at the bottom of the hierarchy).
- Functions evaluated to get parameters for the base constructor end up having a different edge depending on whether they're on inheritance specifier list or modifier list. Note for example that in `F` we have `constructor of E -> fFD` and `constructor of F -> fFE()`. They look like called from different contracts but they're both called for constructor parameters in `F`.
    - This might actually be what @chriseth meant in https://github.com/ethereum/solidity/pull/10332#pullrequestreview-588767032. Changing `m_currentNode` there could change this.

Reply:
This has now changed so I'm marking it as resolved. See https://github.com/ethereum/solidity/pull/10973#discussion_r579244040 for further discussion.

BTW, the test case visible in the snippet above has been renamed to `inheritance_specifiers` and I removed the one that was previously under this name. I think that `diamond_inheritance` covers everything from the old test case and does it better.